phoenix-core/src/main/java/org/apache/phoenix/execute/MutationState.java

```diff
@@ -931,220 +944,323 @@ private void send(Iterator<TableRef> tableRefIterator) throws SQLException {
                             }
-                            if (sqlE != null) { throw sqlE; }
                         }
+                        if (sqlE != null) { throw sqlE; }
                     }
-                } while (shouldRetry && retryCount++ < 1);
+                }
+            } while (shouldRetry && retryCount++ < 1);
+        }
+    }
+
+    private Map<TableInfo, List<Mutation>> filterIndexCheckerMutations(Map<TableInfo, List<Mutation>> mutationMap,
+            Map<TableInfo, List<Mutation>> deletedIndexMutations) {
+        Map<TableInfo, List<Mutation>> verifiedMutationsMap = new LinkedHashMap<>();
+        Iterator<Entry<TableInfo, List<Mutation>>> mapIter = mutationMap.entrySet().iterator();
+        while (mapIter.hasNext()) {
+            Entry<TableInfo, List<Mutation>> pair = mapIter.next();
+            TableInfo tableInfo = pair.getKey();
+            if (IndexUtil.isGlobalIndexCheckerEnabled(connection, tableInfo.hTableName)) {
+                PTable table = tableInfo.getOrigTableRef().getTable();
+                byte[] emptyCF = SchemaUtil.getEmptyColumnFamily(table);
+                byte[] emptyCQ = EncodedColumnsUtil.getEmptyKeyValueInfo(table).getFirst();
+
+                List<Mutation> mutations = pair.getValue();
+
+                if (mutations.size() == 0) {
+                    continue;
+                }
+
+                List<Mutation> indexMutations = new LinkedList<>();
+                for (Mutation m : mutations) {
+                    if (m instanceof Delete) {
+                        Put put = new Put(m.getRow());
+                        put.addColumn(emptyCF, emptyCQ, m.getTimestamp(), FALSE_BYTES);
+                        indexMutations.add(put);
+                        if (deletedIndexMutations.containsKey(tableInfo)) {
+                            deletedIndexMutations.get(tableInfo).add(m);
+                        } else {
+                            List<Mutation> mts = Lists.newArrayList();
+                            mts.add(m);
+                            deletedIndexMutations.put(tableInfo, mts);
+                        }
+                    }
+                }
+
+                verifiedMutationsMap.put(pair.getKey(), indexMutations);
+                mapIter.remove();
+            }
+
+        }
+        return verifiedMutationsMap;
+    }
+
+    private Map<TableInfo, List<Mutation>> setVerifiedColumnForIndexes(List<TableInfo> indexTables, List<Mutation> dataMutations,
+            Map<TableInfo, List<Mutation>> deletedIndexMutations, byte[] verifiedBytes)
+            throws SQLException {
+        Map<TableInfo, List<Mutation>> updatedMap = new LinkedHashMap<>();
+        updatedMap.putAll(deletedIndexMutations);
+        for (TableInfo tableInfo : indexTables) {
+            PTable indexTable = null;
+            PTable dataTable = tableInfo.origTableRef.getTable();
+
+            List<PTable> indexes = dataTable.getIndexes();
```

Comment:
This function will return all the indexes, including local indexes. Is that the expectation?

Reply:
That is fine. It will not match the GlobalIndexChecker index name